### PR TITLE
[WIP] enh(dashboard): add shares to dashboard listing and details

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description

This PR intends to add the contacts and contactgroups shares on the dashboard listing and details

The shares are displayed only if the user is editor.

```
  "result": [
    {
      "id": 1,
      "name": "aaaa",
      "description": "aaaa",
      . . .
      "own_role": "viewer",
      "shares": {
        "contacts": [], // Empty as own role is viewer
        "contact_groups": [] // Empty as own role is viewer
      }
    },
    {
      "id": 2,
      "name": "bbbb",
      "description": "bbbb",
     . . .
      "own_role": "editor",
      "shares": {
        "contacts": [
          {
            "id": 1,
            "name": "admin admin",
            "email": "admin@centreon.com",
            "role": "editor"
          },
          {
            "id": 18,
            "name": "User",
            "email": "user@localhost",
            "role": "viewer"
          },
          {
            "id": 20,
            "name": "test",
            "email": "test",
            "role": "editor"
          }
        ],
        "contact_groups": [
          {
            "id": 6,
            "name": "CG1",
            "role": "viewer"
          }
        ]
      }
    }
  ],
```

**Fixes** # MON-33355

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [x] 23.10.x
- [x] 24.04.x (master)

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/src/Core/Dashboard/Infrastructure/Repository/DbReadDashboardShareRepository.php [hunk 1]
```diff
@@ -234,6 +234,130 @@ public function getMultipleSharingRoles(ContactInterface $contact, Dashboard ...
...
+                $result['contact_name'],
+                $result['contact_email'],
+                $this->stringToRole($result['role'])
+            );
+        }
+
+        return $shares;
+    }
+
+    public function findDashboardsContactGroupShares(Dashboard ...$dashboards): array
```
Comment (jeremyjaouen) [inline centreon/src/Core/Dashboard/Infrastructure/Repository/DbReadDashboardShareRepository.php:297]:
```suggestion
    /**
     * @inheritDoc
     */
    public function findDashboardsContactGroupShares(Dashboard ...$dashboards): array
```

File: centreon/src/Core/Dashboard/Application/UseCase/FindDashboard/FindDashboardFactory.php [hunk 2]
```diff
@@ -69,6 +75,26 @@ public static function createResponse(
...
+        if ($ownRole === DashboardSharingRole::Editor && array_key_exists($dashboard->getId(), $contactGroupShares)) {
+            $response->shares['contact_groups'] = array_map(
+                static fn (DashboardContactGroupShare $contactGroupShare): array => [
+                    'id' => $contactGroupShare->getContactGroupId(),
+                    'name' => $contactGroupShare->getContactGroupName(),
+                    'role' => $contactGroupShare->getRole(),
+                ],
+                $contactGroupShares[$dashboard->getId()]);
+        }
+
```
Comment (adr-mo) [inline centreon/src/Core/Dashboard/Application/UseCase/FindDashboard/FindDashboardFactory.php:97]: You could add a few comments to explicit the functional behaviour wanted for this

**Review by adr-mo — DISMISSED** [review]

File: centreon/src/Core/Dashboard/Application/UseCase/FindDashboard/FindDashboardFactory.php [hunk 3]
```diff
@@ -69,6 +75,26 @@ public static function createResponse(
...
+        if ($ownRole === DashboardSharingRole::Editor && array_key_exists($dashboard->getId(), $contactGroupShares)) {
+            $response->shares['contact_groups'] = array_map(
+                static fn (DashboardContactGroupShare $contactGroupShare): array => [
+                    'id' => $contactGroupShare->getContactGroupId(),
+                    'name' => $contactGroupShare->getContactGroupName(),
+                    'role' => $contactGroupShare->getRole(),
+                ],
+                $contactGroupShares[$dashboard->getId()]);
+        }
```
Comment (jeremyjaouen) [inline centreon/src/Core/Dashboard/Application/UseCase/FindDashboard/FindDashboardFactory.php:96]:
```suggestion
        // Add shares only if the user if editor, as the viewers should not be able to see shares.
        if ($ownRole === DashboardSharingRole::Editor && array_key_exists($dashboard->getId(), $contactShares)) {
            $response->shares['contacts'] = array_map(static fn (DashboardContactShare $contactShare): array => [
                'id' => $contactShare->getContactId(),
                'name' => $contactShare->getContactName(),
                'email' => $contactShare->getContactEmail(),
                'role' => $contactShare->getRole(),
            ],
            $contactShares[$dashboard->getId()]);
        }

        if ($ownRole === DashboardSharingRole::Editor && array_key_exists($dashboard->getId(), $contactGroupShares)) {
            $response->shares['contact_groups'] = array_map(
                static fn (DashboardContactGroupShare $contactGroupShare): array => [
                    'id' => $contactGroupShare->getContactGroupId(),
                    'name' => $contactGroupShare->getContactGroupName(),
                    'role' => $contactGroupShare->getRole(),
                ],
                $contactGroupShares[$dashboard->getId()]);
        }
```

File: centreon/src/Core/Dashboard/Application/UseCase/FindDashboards/FindDashboardsFactory.php [hunk 4]
```diff
@@ -70,6 +76,25 @@ public static function createResponse(
...
+
+            if ($ownRole === DashboardSharingRole::Editor && array_key_exists($dashboard->getId(), $contactGroupShares)) {
+                $dto->shares['contact_groups'] = array_map(
+                    static fn (DashboardContactGroupShare $contactGroupShare): array => [
+                        'id' => $contactGroupShare->getContactGroupId(),
+                        'name' => $contactGroupShare->getContactGroupName(),
+                        'role' => $contactGroupShare->getRole(),
+                    ],
+                    $contactGroupShares[$dashboard->getId()]);
+            }
```
Comment (jeremyjaouen) [inline centreon/src/Core/Dashboard/Application/UseCase/FindDashboards/FindDashboardsFactory.php:97]:
```suggestion

            // Add shares only if the user if editor, as the viewers should not be able to see shares.
            if ($ownRole === DashboardSharingRole::Editor && array_key_exists($dashboard->getId(), $contactShares)) {
                $dto->shares['contacts'] = array_map(static fn (DashboardContactShare $contactShare): array => [
                    'id' => $contactShare->getContactId(),
                    'name' => $contactShare->getContactName(),
                    'email' => $contactShare->getContactEmail(),
                    'role' => $contactShare->getRole(),
                ]
                , $contactShares[$dashboard->getId()]);
            }

            if ($ownRole === DashboardSharingRole::Editor && array_key_exists($dashboard->getId(), $contactGroupShares)) {
                $dto->shares['contact_groups'] = array_map(
                    static fn (DashboardContactGroupShare $contactGroupShare): array => [
                        'id' => $contactGroupShare->getContactGroupId(),
                        'name' => $contactGroupShare->getContactGroupName(),
                        'role' => $contactGroupShare->getRole(),
                    ],
                    $contactGroupShares[$dashboard->getId()]);
            }
```

**Review by adr-mo — APPROVED** [review]

**Review by wtermellil — APPROVED** [review]

**Pull request merged**